### PR TITLE
Exclude pinned tabs when closing active item with shortcut

### PR DESCRIPTION
**Pull request by uncenter**
Closes N/A. Changes based on #19593. 

Replicates VS Code's behavior with pinned tabs; CMD + W (or whatever the keybind is) shouldn't close pinned tabs - a manual context menu action or close button click, or unpin and then close shortcut is needed instead.

I'm not sure when `close_pinned` should be true for the Vim shortcuts so I'd appreciate some guidance there as I probably need to change some of those options. I'm also not sure if I should add tests here.

Release Notes:

- Fixes close tab keyboard shortcut closing pinned tabs.

### CODE REVIEW

File: crates/vim/src/command.rs [hunk 1]
```diff
@@ -510,37 +510,45 @@ fn generate_commands(_: &AppContext) -> Vec<VimCommand> {
...
             ("x", "it"),
             workspace::CloseActiveItem {
                 save_intent: Some(SaveIntent::SaveAll),
+                close_pinned: false,
             },
         )
         .bang(workspace::CloseActiveItem {
             save_intent: Some(SaveIntent::Overwrite),
+            close_pinned: false,
         }),
```
Comment (ConradIrwin) [inline crates/vim/src/command.rs:541]: Thanks for adding this to all the vim commands!

In vim, the "bang" versions should have `close_pinned: true`. (so `:q` doesn't work, but `:q!` does).

Comment (uncenter) [reply crates/vim/src/command.rs:541]: Ah right that makes sense, I didn't read closely enough to realize that's what bang was referring to 😅

**Comment by ConradIrwin** [comment]
@uncenter I think you're missing a check for action.close_pinned in the handler for CloseActiveItem.

**Comment by uncenter** [comment]
Ah you are right, not sure how I missed that! 😄

**Comment by uncenter** [comment]
Huh I tried adding some logic to actually consider close_pinned, and in doing so changed the keymap defaults to have `close_pinned` true (previous behavior) and I can't figure out why it isn't working.

```diff
diff --git a/assets/keymaps/default-linux.json b/assets/keymaps/default-linux.json
index 2c391aff9d..771aa5fb86 100644
--- a/assets/keymaps/default-linux.json
+++ b/assets/keymaps/default-linux.json
@@ -253,7 +253,7 @@
       "ctrl-shift-pagedown": "pane::SwapItemRight",
       "back": "pane::GoBack",
       "forward": "pane::GoForward",
-      "ctrl-w": ["pane::CloseActiveItem", { "close_pinned": false }],
+      "ctrl-w": ["pane::CloseActiveItem", { "close_pinned": true }],
       "ctrl-f4": ["pane::CloseActiveItem", { "close_pinned": false }],
       "alt-ctrl-t": ["pane::CloseInactiveItems", { "close_pinned": false }],
       "alt-ctrl-shift-w": "workspace::CloseInactiveTabsAndPanes",
diff --git a/assets/keymaps/default-macos.json b/assets/keymaps/default-macos.json
index 0f4bbff070..6efab06ad4 100644
--- a/assets/keymaps/default-macos.json
+++ b/assets/keymaps/default-macos.json
@@ -298,7 +298,7 @@
       "alt-cmd-right": "pane::ActivateNextItem",
       "ctrl-shift-pageup": "pane::SwapItemLeft",
       "ctrl-shift-pagedown": "pane::SwapItemRight",
-      "cmd-w": ["pane::CloseActiveItem", { "close_pinned": false }],
+      "cmd-w": ["pane::CloseActiveItem", { "close_pinned": true }],
       "alt-cmd-t": ["pane::CloseInactiveItems", { "close_pinned": false }],
       "ctrl-alt-cmd-w": "workspace::CloseInactiveTabsAndPanes",
       "cmd-k u": ["pane::CloseCleanItems", { "close_pinned": false }],
diff --git a/crates/workspace/src/pane.rs b/crates/workspace/src/pane.rs
index d94c8af034..a999ac773c 100644
--- a/crates/workspace/src/pane.rs
+++ b/crates/workspace/src/pane.rs
@@ -1127,7 +1127,11 @@ impl Pane {
             return None;
         }
         let active_item_id = self.items[self.active_item_index].item_id();
-        if self.is_non_closeable_item(&active_item_id) {
+        if self.is_non_closeable_item(&active_item_id, action.close_pinned) {
+            println!(
+                "is non closeable item, close_pinned: {}",
+                action.close_pinned
+            );
             return None;
         };
         Some(self.close_item_by_id(
@@ -2707,23 +2711,21 @@ impl Pane {
         self.display_nav_history_buttons = display;
     }
 
-    fn is_non_closeable_item(&self, item_id: &EntityId) -> bool {
+    fn is_non_closeable_item(&self, item_id: &EntityId, close_pinned: bool) -> bool {
         if let Some(ix) = self.index_for_item_id(*item_id) {
-            self.is_tab_pinned(ix)
+            println!("index is some for id");
+            !close_pinned && self.is_tab_pinned(ix)
         } else {
+            println!("no index for id");
             true
         }
     }
 
     fn get_non_closeable_item_ids(&self, close_pinned: bool) -> Vec<EntityId> {
-        if close_pinned {
-            return vec![];
-        }
-
         self.items
             .iter()
             .map(|item| item.item_id())
-            .filter(|item_id| self.is_non_closeable_item(item_id))
+            .filter(|item_id| self.is_non_closeable_item(item_id, close_pinned))
             .collect()
     }
```

After running and pressing Cmd + W, looking at the console output I just see:
```
index is some for id
is non closeable item, close_pinned: false
```
Which indicates that close_pinned is still false - despite that being changed in the keymap and having no custom user keymap being set. Struggling to debug this, would appreciate if you can maybe take a look.

I'm also wondering how we can change focus to the previous active pane without actually closing the current pinned one, to fully replicate VS Code here. This here looks related but I'm not sure if this needs to be split off into a function so we can reuse the logic or if something else entirely is needed:

https://github.com/zed-industries/zed/blob/7330b5c5e7cb2081236964c256ac2b8d63d5bb12/crates/workspace/src/pane.rs#L1468-L1488

**Comment by osiewicz** [comment]
A wild guess is that `serde(default)` on `close_pinned` is somehow messing with deserialization?

**Comment by mikayla-maki** [comment]
Sounds like there's some issues with this PR, I'm going to mark it as draft until they're resolved :)

**Comment by uncenter** [comment]
Still can't figure out why it isn't working, going to raise an issue instead and hope someone else can do it.

**Pull request closed without merging**